feat(bin, engine, prune): spawn pruning task from the engine

shekhirin opened the pull request:
Ref https://github.com/paradigmxyz/reth/issues/3439

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -360,6 +360,14 @@ impl Command {
             None
         };
 
+        let pruner = config.prune.map(|prune_config| {
+            reth_prune::Pruner::new(
+                blockchain_db.canonical_state_stream(),
+                prune_config.block_interval,
+                tree_config.max_reorg_depth(),
+            )
+        });
```
should we have an info log for when this is installed?

mattsse commented on crates/consensus/beacon/src/engine/metrics.rs:
```diff
@@ -13,6 +13,8 @@ pub(crate) struct EngineMetrics {
     pub(crate) forkchoice_updated_messages: Counter,
     /// The total count of new payload messages received.
     pub(crate) new_payload_messages: Counter,
+    /// The number of times the pruner was run.
+    pub(crate) pruner_runs: Counter,
```
do we also want `pruned_blocks`?

shekhirin replied:
I think it will need to be a separate detailed prune metrics struct, as we can prune different number of blocks for different prune parts.

mattsse commented on crates/consensus/beacon/src/engine/mod.rs:
```diff
@@ -213,7 +219,7 @@ impl BeaconConsensusEngineHandle {
 /// If the future is polled more than once. Leads to undefined state.
 #[must_use = "Future does nothing unless polled"]
-pub struct BeaconConsensusEngine<DB, BT, Client>
+pub struct BeaconConsensusEngine<DB, BT, Client, St>
```
I wanted to move the CanonStateNotification out of the blockchain tree anyway and let the beacon engine do the notifying.

So we can keep this, otherwise I'd prefer if the Pruner could just box the stream, perhaps it should anyways? just to make it easier to use, the type only fires an event every 12s so there's no perf to gain/lose with dyn

mattsse commented on crates/consensus/beacon/src/engine/prune.rs:
```diff
@@ -0,0 +1,160 @@
+/// This type controls the [Pruner].
+pub(crate) struct EnginePruneController<St> {
+    /// The current state of the pruner.
+    pruner_state: PrunerState<St>,
```
why can't the pruner just live on its own task?

because we need tight control over when it's running and when not?

shekhirin replied:
yes, correct. Otherwise we won't be able to control who owns the write tx, OR we will have a more complicated flow like that:
1. Pruner is notified with `CanonStateNotification` by Blockchain Tree
2. Pruner notifies Sync Controller that it wants to start the pruning process with `RequestPruneStart`
3. Sync Controller processes the `RequestPruneStart` message, sets  `self.is_pruning = true` disallowing the Engine to process any incoming FCUs, replies with `ResponsePruneStart`
4. Pruner starts the pruning process and sends `PruneFinished` when finishes
5. Sync Controller processes the `PruneFinished` message, sets `self.is_Pruning = false` allowing the Engine to continue processing incoming FCUs

mattsse replied:
> yes, correct. Otherwise we won't be able to control who owns the write tx

I see, yeah that makes sense

mattsse replied:
> Pruner is notified with CanonStateNotification by Blockchain Tree

hmm, if the beacon engine controls this, then it doesn't need it? I think that's my main confusion, why does it even need this when the pruner is under the control of the engine and it only needs to know the local head

shekhirin replied:
> hmm, if the beacon engine controls this, then it doesn't need it?

ah, yes, that totally makes sense. We can just pass the local head from the engine.
But I would still prefer to leave the `Pruner::check_tip` method, so we would have the logic of minimum pruning interval inside `Pruner`. It will just accept the `tip_block_number` from the caller now.

mattsse replied:
so we have

`BeaconEngine` polls `PruneController` with head number, `PrunerController` can spawn a task to prune depending on the head number
`BeaconEngine` needs to prevent commits while pruner is active

shekhirin replied:
correct. So there needs to be a way to communicate to `BeaconEngine` that the pruner is currently running, and I chose the same way as with pipeline: have a separate controller which spawns the pruner and contains the current state (idle/running).

mattsse commented on crates/prune/src/pruner.rs:
```diff
@@ -0,0 +1,168 @@
+pub struct Pruner<St> {
+    /// Stream of canonical state notifications. Pruning is triggered by new incoming
+    /// notifications.
+    canon_state_stream: St,
```
maybe we can make this boxed?
the generics are a bit painful for this

shekhirin replied:
right, agree

mattsse commented on crates/prune/src/pruner.rs:
```diff
@@ -0,0 +1,168 @@
+};
+use tracing::debug;
+
+/// The future that returns the owned pipeline and the result of the pipeline run. See
+/// [Pruner::run_as_fut].
+pub type PrunerFut<St> = Pin<Box<dyn Future<Output = PrunerWithResult<St>> + Send>>;
```
not a fan of this pattern, we had to do it for the pipeline because of its blocking behavior which makes it not pollable.

But I don't really understand why we need to spawn and stop the pruner task?

hmm, if we spawn this, how long does it run then?

shekhirin replied:
> But I don't really understand why we need to spawn and stop the pruner task?

1. It needs to acquire the write tx that shouldn't be shared by any other tasks, so we need one place to control who owns the write tx.
2. It's a blocking task, same as pipeline because it interacts with database.

rkrasiuk commented on crates/prune/src/pruner.rs:
```diff
@@ -0,0 +1,168 @@
+    pub fn check_tip(&mut self, cx: &mut Context<'_>) -> Option<BlockNumber> {
+        let mut latest_canon_state = None;
+        while let Poll::Ready(Some(canon_state)) =
+            Pin::new(&mut self.canon_state_stream).poll_next(cx)
```
draining seems right, but it should already be pinned. see [`maintain_transaction_pool`](https://github.com/paradigmxyz/reth/blob/8416dedeb8e499f5764f0b901efb8f166d089fbf/crates/transaction-pool/src/maintain.rs#L46) for sample usage

mattsse replied:
hmm, the tokio::broadcast has a `latest` function I believe, we could expose this but need to add a custom trait I think.
but as commented, maybe we don't need this at all, depending on how the task should work

mattsse reviewed (changes requested):
> Pruner::check_tip drains the stream of CanonStateNotification, checking the latest available tip block number. If there's any and it passes the minimum pruning interval, returns Some(_), otherwise None.

can you elaborate on what this pruner task does with these events or how the task works in detail? because I'm a bit confused by the check_tip, if it only checks the tip, then it doesn't need the stream of new blocks?
This question is related to how the pruner task is controlled I believe

> When pruning finishes, EnginePruneController sets its pruner_state back to Idle, allowing BeaconConsensusEngine to process any incoming FCUs again.

why do we need to do this? why can't the pruner just live on its own? because blocking?

I personally very much dislike the pattern where we need to spawn the object and then receive it back eventually. we had to do this for the pipeline because we made all of its stages async.
But if we spawn prune jobs based on block number, does it even need to be async?

> If EnginePruneController received the tip block number, it spawns the pruning task in a separate thread

why does the pruner need the block stream then?

shekhirin commented:
@mattsse 

> because I'm a bit confused by the check_tip, if it only checks the tip, then it doesn't need the stream of new blocks?

this is correct now, yes. We've switched from a background always-running task to a spawnable from engine, so we know the tip block number now, and can pass it to the pruning task. Fixed.

> why do we need to do this? why can't the pruner just live on its own? because blocking?

yes, the pruning is blocking and the workload is similar to pipeline's as it will do a lot of db writes.

> I personally very much dislike the pattern where we need to spawn the object and then receive it back eventually. we had to do this for the pipeline because we made all of its stages async.

`Pruner::run` needs to take a mutable self, so we need to give the blocking task exclusive mutable reference of `Pruner`, and such approach with `Idle(Option<Pruner>)` looks like a good solution for me. Also, that way we would have it similar to how sync controller spawns the pipeline, hence it simplifies the understanding of the codebase.

> But if we spawn prune jobs based on block number, does it even need to be async?

it doesn't seem now, so I made `Pruner::run` sync.

mattsse commented on crates/consensus/beacon/src/engine/mod.rs:
```diff
@@ -852,10 +874,13 @@ where
+            // because they hold exclusive access to the database
             self.try_insert_new_payload(block)
+        } else if self.is_prune_active() {
+            warn!(target: "consensus::engine", "Pruning is in progress, buffering new payload.");
```
should this be a warning?

I think we should keep the else branch and put the log in an if, so we only have one `self.try_buffer_payload(block)`

shekhirin replied:
I think we should have this as a warning, yes, so the node operator could see that they should probably lower the pruning interval. It's not as critical as during the FCU, but still thought worth warning about.

Made one else branch, agree.

mattsse commented on crates/consensus/beacon/src/engine/mod.rs:
```diff
@@ -1407,10 +1473,24 @@ where
+            }
+
+            // check prune events if pipeline is idle and both engine and sync events are pending
+            if this.sync.is_pipeline_idle() && engine_messages_pending & sync_pending {
```
why can we only do pruning if the others are pending?

shekhirin replied:
the logic is to prioritize processing engine & sync messages by processing them fully first, only then do pruning

shekhirin replied:
hmm, I think we can improve this condition in such way:
```suggestion
            if this.sync.is_pipeline_idle() && (this.is_prune_active() || engine_messages_pending & sync_pending) {
```

So, check pruning if:
1. Pipeline is idle
2. Either pruning is running and we need to prioritize checking its events OR no engine and sync messages are pending and we may start pruning

mattsse replied:
right, yep that works

mattsse commented on crates/consensus/beacon/src/engine/mod.rs:
```diff
@@ -1407,10 +1473,24 @@ where
+
+            // check prune events if pipeline is idle and both engine and sync events are pending
+            if this.sync.is_pipeline_idle() && engine_messages_pending & sync_pending {
+                if let Some(prune) = &mut this.prune {
```
```suggestion
                if let Some(ref mut prune) = this.prune {
```

mattsse commented on crates/prune/src/error.rs:
```diff
@@ -0,0 +1,4 @@
+use thiserror::Error;
+
```
this is a placeholder for now?

mattsse reviewed (commented):
just some questions re poll logic

shekhirin commented on crates/consensus/beacon/src/engine/prune.rs:
```diff
@@ -0,0 +1,152 @@
+                // Can not make any progress
+                return Poll::Pending
+            }
+        }
```
I think not, removed. If pruner didn't spawn and no progress was made in the running pruner, return `Pending`.

mattsse reviewed (approved):
lgtm

mattsse commented on crates/consensus/beacon/src/engine/prune.rs:
```diff
@@ -0,0 +1,149 @@
+        }
+
+        // Poll pruner and check its status
+        if let Poll::Ready(event) = self.poll_pruner(cx) {
+            return Poll::Ready(event)
+        }
```
nit: we can return `self.poll_pruner(cx)` directly?

joshieDo reviewed (approved):
lgtm

in the future (wrt snapshots), we might want to extend and add another event like `OnHold`, when we need to prune in chunks using checkpoints

mattsse commented on crates/consensus/beacon/src/engine/mod.rs:
```diff
@@ -1018,21 +1043,26 @@ where
+    ///
+    /// This is invoked after a pruner run to update the tree with the most recent canonical
+    /// hashes.
+    fn update_tree_on_finished_pruner(&mut self) -> Result<(), Error> {
+        self.blockchain.restore_canonical_hashes()
+    }
```
I like having this extra function so we can extend this if we want to call finaluze_block eventually

mattsse reviewed (approved):
lgtm

shekhirin commented:
finally merging 🫡
